### ISSUE TITLE
TypedDict union not narrowing correctly

### ISSUE THREAD

**Issue by mwaddoups** [microsoft/pyright#5436]
**Describe the bug**

When you union two typed dicts with different fields, asserting that a field exists won't let you access that field.

Related to the point I raised https://github.com/microsoft/pyright/discussions/5435 - but this is a more clear issue.

**To Reproduce**
```python
from typing import TypedDict, Literal

class A(TypedDict): 
    name: Literal["A"]
    foo: str

class BC(TypedDict):
    name: Literal["B", "C"]
    bar: str

AorBC = A | BC

def some_func(x: AorBC):
    assert 'bar' in x

    y = x['bar']  # this has "bar" is not a defined key in "A"
```

**Expected behavior**
Type of `x` should narrow to `BC` following the assertion, allowing access to `'bar'`.

I can understand that this is maybe not common usage of TypedDict's so may not be a supported narrowing case - but I think it's very unexpected to have a type checker complain after you've asserted that the field is present when you try to access it.

**Comment by erictraut** [comment]
It is not safe for pyright to perform narrowing in this case becuase a TypedDict can be subclassed. For example, a subclass of `A` might add a field `bar`. If you mark your two TypedDict definitions as `@final`, pyright will perform type narrowing as you are hoping.

I'll note that the use of `@final` with `TypedDict` is a bit sketchy because `TypedDict` is a structural type (like a protocol), not a nominal type. There are some active discussions in the typing community about whether `@final` should be applicable to `TypedDict`. Depending on the resolution of those discussions, I might remove support for this type narrowing altogether. Note that mypy doesn't currently support this narrowing pattern even if the TypedDict is marked `@final`.
